### DISCUSSION TITLE
Guidance for using JoinableTaskFactory in DotNet application?

### DISCUSSION THREAD

**Discussion by spearsandshields** [microsoft/vs-threading#1236] [Q&A]
Hi, I am planning to use JoinableTaskFactory from this library to address sync-over-async anti-patterns in a existing codebase to prevent deadlocks. 

I am still a bit confused regarding the JoinableTaskContext, and would appreciate some guidance. Should it be instantiated for each request, or should it be assigned to a specific thread from the threadpool?  (since I believe there isn't technically a UI/main thread in DotNet apps)
```
public class JoinableTaskFactoryUtil {
    private static readonly System.Threading.AsyncLocal<JoinableTaskContext> _context =
        new System.Threading.AsyncLocal<JoinableTaskContext>();

    public static JoinableTaskFactory Factory =>
        _context.Value?.Factory ?? (_context.Value = new JoinableTaskContext()).Factory;
}
```
Due to the above confusion, I went with the above pattern as a provider for the JoinableTaskContext. 

However as I ran through a few benchmark tests, the results seemed a bit odd.  Looks like for the nested cases, the JTF actually results in degraded performance. (While i think the chance of using nested factory is low, the possibility cannot be ruled out)
Could anyone provided some clarification here. Thanks in advance

| Method                                      | Mean        | Error     | StdDev    | Gen0      | Allocated |
|-------------------------------------------- |------------:|----------:|----------:|----------:|----------:|
| AsyncAllTheWay                              |   109.55 ms |  0.911 ms |  0.807 ms |         - |   46.4 KB |
| WithSingleJTF                               |   108.97 ms |  0.659 ms |  0.514 ms |         - |     56 KB |
| WithJTFMultipleContextNested                | 3,301.22 ms | 21.665 ms | 19.206 ms |         - |    176 KB |
| WithJTFAndFactoryUtilNested                 | 3,291.48 ms | 21.336 ms | 19.957 ms |         - |    168 KB |
| SyncOverAsync1                              |   108.44 ms |  0.660 ms |  1.173 ms | 1000.0000 | 689.26 KB |
| SyncOverAsync2                              |   108.03 ms |  1.904 ms |  1.590 ms |         - |    264 KB |
| WithJTFDeadlockAvoidance                    |    14.44 ms |  0.289 ms |  0.729 ms |         - |    248 KB |
| SyncOverAsyncWithDeadlockPotentialRewritten |   509.78 ms |  5.616 ms |  5.253 ms |         - |    208 KB |

Below I will attach the BenchmarkCode

```
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using Microsoft.VisualStudio.Threading;
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

public class JoinableTaskFactoryUtil {
    private static readonly System.Threading.AsyncLocal<JoinableTaskContext> _context =
        new System.Threading.AsyncLocal<JoinableTaskContext>();

    public static JoinableTaskFactory Factory =>
        _context.Value?.Factory ?? (_context.Value = new JoinableTaskContext()).Factory;
}

[MemoryDiagnoser]
public class JTFBenchmark {
    private const int Timeout = 500; // Timeout to prevent indefinite blocking.
    private const int ConcurrentTasksCount = 30; // Number of concurrent tasks.


    private async Task RunWithCustomSyncContext(Func<Task> func) {
        var prevCtx = SynchronizationContext.Current;
        try {
            var ctx = new CustomSynchronizationContext();
            SynchronizationContext.SetSynchronizationContext(ctx);
            await func();
        }
        finally {
            SynchronizationContext.SetSynchronizationContext(prevCtx); // Restore the previous context.
        }
    }


    private async Task NestedTask() {
        await Task.Delay(100);
    }

    private async Task ConcurrentTask(Func<Task> taskMethod) {
        var tasks = Enumerable.Range(0, ConcurrentTasksCount).Select(_ => taskMethod()).ToArray();
        await Task.WhenAll(tasks);
    }

    [Benchmark]
    public async Task AsyncAllTheWay() {
        await RunWithCustomSyncContext(async () => { await ConcurrentTask(async () => { await NestedTask(); }); });
    }
    
    [Benchmark]
    public async Task WithSingleJTF() {
        await RunWithCustomSyncContext(async () => {
            JoinableTaskFactoryUtil.Factory.Run(async () => {
                await ConcurrentTask(async () => { await NestedTask(); });
            });
        });
    }

    [Benchmark]
    public async Task WithJTFMultipleContextNested() {
        await RunWithCustomSyncContext(async () => {
            await ConcurrentTask(async () => {
                var context = new JoinableTaskContext();
                var joinableTask = context.Factory.RunAsync(async () => {
                    var context2 = new JoinableTaskContext();
                    await context2.Factory.RunAsync(async () => { await NestedTask(); });
                });
                joinableTask.Join();
            });
        });
    }
    
    
    [Benchmark]
    public async Task WithJTFAndFactoryUtilNested() {
        await RunWithCustomSyncContext(async () => {
            await ConcurrentTask(async () => {
                var joinableTask = JoinableTaskFactoryUtil.Factory.RunAsync(async () => {
                    await JoinableTaskFactoryUtil.Factory.RunAsync(async () => { await NestedTask(); });
                });
                joinableTask.Join();
            });
        });
    }

    [Benchmark]
    public void SyncOverAsync1() {
        RunWithCustomSyncContext(() => {
            var tasks = Enumerable.Range(0, ConcurrentTasksCount).Select(_ => Task.Run(() => {
                Task.Run(async () => { Task.Run(async () => { await NestedTask(); }).Wait(); }).Wait();
            })).ToArray();

            Task.WaitAll(tasks); // This blocks synchronously until all tasks have completed.
            return Task.CompletedTask;
        });
    }

    [Benchmark]
    public Task SyncOverAsync2() {
        return RunWithCustomSyncContext(async () => {
            var tasks = Enumerable.Range(0, ConcurrentTasksCount).Select(_ => Task.Run(async () => {
                Task taskInner = Task.Run(async () => { await NestedTask(); });
                Task.WaitAll(taskInner); 
                return Task.CompletedTask;
            })).ToArray();
            Task.WaitAll(tasks); 
        });
    }
    
    [Benchmark]
    public void WithJTFDeadlockAvoidance() {
        RunWithCustomSyncContext(() => {
            Parallel.For(0, ConcurrentTasksCount, _ => {
                // Here, the current synchronization context is already the custom one set by RunWithCustomSyncContext.
    
                // Use JoinableTaskFactory to run and synchronously wait for the async operation.
                JoinableTaskFactoryUtil.Factory.Run(async delegate {
                    await Task.Delay(10).ConfigureAwait(true); // Try to marshal back to the original context.
                });
            });
    
            return Task.CompletedTask;
        });
    }
    
    
    [Benchmark]
    public void SyncOverAsyncWithDeadlockPotentialRewritten() {
        RunWithCustomSyncContext(() => {
            Parallel.For(0, ConcurrentTasksCount, _ => {
                // Here, the current synchronization context is already the custom one set by RunWithCustomSyncContext.
    
                var outerTask = Task.Run(async () => {
                    await Task.Delay(10).ConfigureAwait(true); // Try to marshal back to the original context.
                });
    
                // This will likely deadlock as the outerTask tries to marshal back to the custom context,
                // but the custom context is blocked waiting for outerTask to complete.
                if (!outerTask.Wait(Timeout)) {
                    throw new TimeoutException("Potential deadlock detected.");
                }
            });
            return Task.CompletedTask;
        });
    }

    private class CustomSynchronizationContext : SynchronizationContext {
        public override void Post(SendOrPostCallback d, object state) {
            // Simply queue the delegate and don't execute it immediately.
            ThreadPool.QueueUserWorkItem(_ => d(state));
        }
    }

    class Program {
        static void Main(string[] args) {
            var summary = BenchmarkRunner.Run<JTFBenchmark>();
        }
    }
}
```

**Comment by AArnott** [comment]
You didn't mention what kind of application you're building. 

> since I believe there isn't technically a UI/main thread in DotNet apps

This is incorrect. .NET apps are no different from .NET Framework or any other app platform in this respect. Every app has an entrypoint thread, and most or all apps with a GUI have a UI thread, which tends to be the entrypoint thread, or what we typically call a main thread. 
.NET WPF/WinForms apps for example absolutely have a main/UI thread.
ASP.NET as a backend application does not.

If your app has a main thread, you should first create a single-threaded SynchronizationContext on it and then create _one_ JoinableTaskContext on that. 
If your app doesn't have a main thread (as you seem to allude to), you should create _one_ JoinableTaskContext on a thread where `SynchronizationContext.Current == null`. You can store that in a static field somewhere and use its `Factory` method everywhere.

You should _not_ ever have more than one `JoinableTaskContext` instance in your process unless you have more than one main thread. And heaven help you if you do. 😬

**Reply by spearsandshields** [reply]
Thanks for the insights, AArnott. I am sorry for not articulating it well. I meant to say DotNet WebAPI , basically an app using DotNet web framework, such as ASP.NET, or ASP.NET core WebApi (It somehow targets both) hosted on something like IIS or equivalent.  My goal was to detect and fix deadlock prone code using the analyzer, and vs-threading library. 

**Switch over to async all the way**, or **rewrite it with JoinableTaskFactory** were the 2 options offered by the analyzer when it encountered sync-over-async code. Since making the violations async-all-the-way could be very costly, and affect a good portion of the code base, i thought JoinableTaskFactory was a more practical approach in the meantime to allow progressive transition to a fully async world. 

This appeared to be a great library to solve our problem, and it has the potential to minimize the chance of getting deadlocks. 
I guess I would appreciate some guidance on the web application best practice regarding this JoinableTaskContext, and whether a static JoinableTaskContext is okay (as i definitely do not want thread congestions, if all requests on relying on this context, which is bound to a specific thread, it would also look simpler than the current ```static AsyncLocal<JoinableTaskContext>``` i showed in the original post!) 

Thank you, and excuse me for my ignorance over this topic!

**Reply by AArnott** [reply]
Ya, do _not_ store JoinableTaskContext in an AsyncLocal. 
I think JTF is likely a good option for you to gradually move sync to async. I don't expect it to protect you from deadlocks on ASP.NET Core. ASP.NET maybe, since it _does_ create a SyncContext that disallows more than one thread per web request.
What JTF brings to the table for you mostly I suspect is that when a sync method calls an async one, its continuations are more likely to execute on the thread that's already blocking anyway rather than consume a second thread. So I would expect overall request throughput to go up. But you should always measure to confirm.
